### PR TITLE
Remove shard ID and unstable suffix when comparing failed job names with the base commit

### LINKED ISSUE #4328: DrCI fails to identify test as broken on trunk
From https://github.com/pytorch/pytorch/pull/104214 DrCI identifies `test_sparse_semi_structured` as NEW failure, even though it is caused by trunk failures introduced by https://github.com/pytorch/pytorch/commit/cf5262a84f815c1e574883bc244333d0d211c7a2 that was later reverted

Hud link: https://hud.pytorch.org/pytorch/pytorch/pull/104214?sha=dd52521aa51e4d609fb97b6a4cf657fc89c49308

image
image

### PR DESCRIPTION
**Pull request by huydhn**
Fixes https://github.com/pytorch/test-infra/issues/4328

This goes together with https://github.com/pytorch/test-infra/pull/4353 and it updates `trymerge` to remove shard ID and the `unstable` suffix when comparing failed job names with the base commit.

### Testing

Add unit tests with the reported issue as a test case https://github.com/pytorch/pytorch/pull/104214 to make sure that the failure there is reported as ignorable broken trunk instead of a new failure.

cc @albanD

### CODE REVIEW

File: .github/scripts/test_trymerge.py [hunk 1]
```diff
@@ -490,6 +492,137 @@ def test_pr_changed_submodule_detection(self, *args: Any) -> None:
         self.assertEqual(pr.get_changed_submodules(), ["third_party/kineto"])
         self.assertFalse(pr.has_invalid_submodule_updates())
 
+    def test_remove_job_name_suffix(self, *args: Any) -> None:
```
Comment (malfet) [inline .github/scripts/test_trymerge.py:495]: You don't really need args here, do you?
```suggestion
    def test_remove_job_name_suffix(self) -> None:
```

File: .github/scripts/trymerge.py [hunk 2]
```diff
@@ -1486,24 +1487,53 @@ def get_rockset_results(head_sha: str, merge_base: str) -> List[Dict[str, Any]]:
         return []
 
 
+REMOVE_JOB_NAME_SUFFIX_REGEX = re.compile(r", [0-9]+, [0-9]+, .+\)")
```
Comment (malfet) [inline .github/scripts/trymerge.py:1490]: NIT: Make sure that this brace is at the end of the statement, and perhaps specify that regex includes the entire suffix 
```suggestion
REMOVE_JOB_NAME_SUFFIX_REGEX = re.compile(r"\(\w+ , [0-9]+, [0-9]+, .+\)$")
```

File: .github/scripts/trymerge.py [hunk 3]
```diff
@@ -1486,24 +1487,53 @@ def get_rockset_results(head_sha: str, merge_base: str) -> List[Dict[str, Any]]:
...
+        return False
+
+    for base_job in base_jobs.values():
+        if (
+            head_job["conclusion"] == base_job["conclusion"]
+            and head_job["failure_captures"] == base_job["failure_captures"]
+        ):
+            return True
+
+    return False
```
Comment (malfet) [inline .github/scripts/trymerge.py:1510]: Nit
```suggestion
    return any(
            head_job["conclusion"] == base_job["conclusion"]
            and head_job["failure_captures"] == base_job["failure_captures"]
            for base_job in base_jobs.values()
        )
```

**Review by malfet — APPROVED** [review]

File: .github/scripts/test_trymerge.py (hunk 1 above)
Comment (huydhn) [reply .github/scripts/test_trymerge.py:495]: Ah, this function doesn't need to use `arg*` but still needs to declare it as the test class itself has 3 `mock.patch`

```
@mock.patch("trymerge.read_flaky_rules", side_effect=empty_flaky_rules)
@mock.patch("trymerge.get_rockset_results", side_effect=empty_rockset_results)
@mock.patch("trymerge.gh_graphql", side_effect=mocked_gh_graphql)
class TestTryMerge(TestCase):
    ...
```

So I think all the test cases need at least 4 arguments.

**Pull request closed without merging**